mautic/mautic
# NPM dependencies - the first taste

| Q                                      | A
| -------------------------------------- | ---
| Bug fix? (use the a.b branch)          | [ ]
| New feature/enhancement? (use the a.x branch)      | [enhancement]
| Deprecations?                          | [ ]
| BC breaks? (use the c.x branch)        | [ ]
| Automated tests included?              | [ ] 
| Related user documentation PR URL      | mautic/mautic-documentation#... 
| Related developer documentation PR URL | mautic/developer-documentation#... 
| Issue(s) addressed                     | Fixes #... 



#### Description:

This PR was inspired by https://github.com/mautic/mautic/pull/10348 and https://github.com/mautic/mautic/pull/10596

The goal where we want to get is to use updated JS dependencies in Mautic 5. It's important to update JS dependencies the same way we update PHP dependencies. Mostly to get all security patches but also new features, bug fixes and performance improvements.

However, all JS dependencies are ["hard-coded" in the Mautic's core](https://github.com/mautic/mautic/tree/5.x/app/bundles/CoreBundle/Assets/js/libraries). Nowadays, there are package managers for JS dependencies like Composer for PHP. I chose to use [NPM](https://www.npmjs.com/) as it is one of the 2 standard package managers out there.

In this PR i'm just tasting the waters of NPM with 2 JS libraries and awaiting feedback. I [installed](https://github.com/mautic/mautic/commit/da3906de220782fef460df6830ab5a89c52f676a) them with these 2 commands:
```
npm install mousetrap@1.4.6
npm install jquery@3.3.1
```
The important part is that I requested to install the same versions of those 2 libraries as we already have in Mautic. There will be another PR where we start upgrading the libraries. Baby steps. This PR is just to discuss if this is the way we want to move forward with the package manager.

Another part of this PR is that there is an new array in the [`AssetGenerationHelper`](https://github.com/mautic/mautic/commit/14b4d729992c96c64ad148c18c157ad640e04ee8) where we can define the JS dependencies that we want to load in the Mautic's administration. This allows us to use the libraries from the `node_modules` folder instead of Mautic's code base. The `node_modules` folder is the same for NPM as the `vendor` folder is for Composer. It's auto-generated folder with the required dependencies.

To make the installation of the NPM packages automated, I [added](https://github.com/mautic/mautic/commit/52872e7651634c2a5db4306c6f127f029395fcfb) the `npm install` command to the Composer's post-install commands. So by running the `composer install` command it will also install JS dependencies.

After that we can safely [remove](https://github.com/mautic/mautic/commit/988be73fff72ea32d0bd8f86d0d938da277cb949) the 2 JS libraries from the Mautic's codebase.

And finally, we must [stop excluding](https://github.com/mautic/mautic/pull/12234/commits/db0d6f65ac67f5e3b48c75fd9fc0bb6d887af901) node_modules folder when building the production Mautic package.

#### Next steps

The next steps will be to 
1. move as many of the existing JS libraries from Mautic's core to NPM dependencies using the same versions if possible.
2. upgrade the libraries to latest versions
3. Ideally configure [Symfony Encore](https://symfony.com/doc/5.4/frontend.html)

#### Steps to test this PR:


1. Open this PR on Gitpod or pull down for testing locally (see docs on testing PRs [here](https://contribute.mautic.org/contributing-to-mautic/tester))
4. Make sure you can navigate the administration as before. That will test if jQuery is present.
5. Make sure that [keyboard shortcuts](https://docs.mautic.org/en/tips-and-tricks) still work. That will test that the Mousetrap library is loaded correctly.
6. Check that there are no JS errors in the browser's dev tools console (skip this if you don't know what it is)
